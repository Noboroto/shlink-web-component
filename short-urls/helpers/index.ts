import { isNil } from 'ramda';
import type { OptionalString } from '../../utils/helpers';
import type { ShortUrlCreationSettings } from '../../utils/settings';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import type { ShortUrl, ShortUrlData } from '../data';

export const shortUrlMatches = (shortUrl: ShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlDataFromShortUrl = (shortUrl?: ShortUrl, settings?: ShortUrlCreationSettings): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls,
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
