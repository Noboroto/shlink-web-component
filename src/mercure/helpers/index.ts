import { EventSourcePolyfill as EventSource } from 'event-source-polyfill';
import type { MercureInfo } from '../reducers/mercureInfo';

export const bindToMercureTopic = <T>(
  mercureInfo: MercureInfo,
  topics: string[],
  onMessage: (message: T) => void,
  onTokenExpired: () => void
) => {
  const { mercureHubUrl, token, loading, error } = mercureInfo;

  if (loading || error || !mercureHubUrl) {
    return undefined;
  }

  const onEventSourceMessage = ({ data }: { data: string }) =>
    onMessage(JSON.parse(data) as T);
  const onEventSourceError = ({ status }: { status: number }) =>
    status === 401 && onTokenExpired();

  const subscriptions = topics.map((topic) => {
    const hubUrl = new URL(mercureHubUrl);

    hubUrl.searchParams.append('topic', topic);
    const es = new EventSource(hubUrl, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    es.onmessage = onEventSourceMessage;
    es.onerror = onEventSourceError;

    return es;
  });

  return () => subscriptions.forEach((es) => es.close());
};
