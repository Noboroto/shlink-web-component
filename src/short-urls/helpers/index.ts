import { isNil } from 'ramda';
import type { ShlinkShortUrl } from '../../api-contract';
import type { OptionalString } from '../../utils/helpers';
import type { ShortUrlCreationSettings } from '../../utils/settings';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import type { ShortUrlData } from '../data';

export const shortUrlMatches = (shortUrl: ShlinkShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShlinkShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlDataFromShortUrl = (
  shortUrl?: ShlinkShortUrl,
  settings?: ShortUrlCreationSettings,
): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls && {
      android: shortUrl.deviceLongUrls.android ?? undefined,
      ios: shortUrl.deviceLongUrls.ios ?? undefined,
      desktop: shortUrl.deviceLongUrls.desktop ?? undefined,
    },
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
