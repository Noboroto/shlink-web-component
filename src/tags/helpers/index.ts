const ONE_OR_MORE_SPACES_REGEX = /\s+/g;

export const normalizeTag = (tag: string) => tag.trim().toLowerCase().replace(ONE_OR_MORE_SPACES_REGEX, '-');
