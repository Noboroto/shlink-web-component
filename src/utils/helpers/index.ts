import { isEmpty, isNil, pipe, range } from 'ramda';
import type { SyntheticEvent } from 'react';

export type OptionalString = string | null | undefined;

export const handleEventPreventingDefault = <T>(handler: () => T) => pipe(
  (e: SyntheticEvent) => e.preventDefault(),
  handler,
);

export const rangeOf = <T>(size: number, mappingFn: (value: number) => T, startAt = 1): T[] =>
  range(startAt, size + 1).map(mappingFn);

export type Empty = null | undefined | '' | never[];

export const hasValue = <T>(value: T | Empty): value is T => !isNil(value) && !isEmpty(value);

export const nonEmptyValueOrNull = <T>(value: T): T | null => (isEmpty(value) ? null : value);

export type BooleanString = 'true' | 'false';

export const parseBooleanToString = (value: boolean): BooleanString => (value ? 'true' : 'false');

export const parseOptionalBooleanToString = (value?: boolean): BooleanString | undefined => (
  value === undefined ? undefined : parseBooleanToString(value)
);
