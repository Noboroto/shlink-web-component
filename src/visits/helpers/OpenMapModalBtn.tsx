import { faMapMarkedAlt as mapIcon } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useDomId, useToggle } from '@shlinkio/shlink-frontend-kit';
import { useCallback, useState } from 'react';
import { Button, Dropdown, DropdownItem, DropdownMenu, DropdownToggle, UncontrolledTooltip } from 'reactstrap';
import type { CityStats } from '../types';
import { MapModal } from './MapModal';
import './OpenMapModalBtn.scss';

interface OpenMapModalBtnProps {
  modalTitle: string;
  activeCities?: string[];
  locations?: CityStats[];
}

export const OpenMapModalBtn = ({ modalTitle, activeCities, locations = [] }: OpenMapModalBtnProps) => {
  const [mapIsOpened, , openMap, closeMap] = useToggle();
  const [dropdownIsOpened, toggleDropdown] = useToggle();
  const [locationsToShow, setLocationsToShow] = useState<CityStats[]>([]);
  const id = useDomId();

  const openMapWithCities = useCallback((cities: CityStats[] = locations) => {
    setLocationsToShow(cities);
    openMap();
  }, [locations]);

  return (
    <>
      {!activeCities && (
        <Button
          color="link"
          className="open-map-modal-btn__btn"
          id={id}
          onClick={() => openMapWithCities()}
        >
          <FontAwesomeIcon icon={mapIcon} />
        </Button>
      )}
      {activeCities && (
        <Dropdown isOpen={dropdownIsOpened} toggle={toggleDropdown}>
          <DropdownToggle color="link" className="open-map-modal-btn__btn" id={id}>
            <FontAwesomeIcon icon={mapIcon} />
          </DropdownToggle>
          <DropdownMenu end>
            <DropdownItem onClick={() => openMapWithCities()}>Show all locations</DropdownItem>
            <DropdownItem
              onClick={() => openMapWithCities(locations.filter(({ cityName }) => activeCities.includes(cityName)))}
            >
              Show locations in current page
            </DropdownItem>
          </DropdownMenu>
        </Dropdown>
      )}
      <UncontrolledTooltip placement="left" target={id}>Show in map</UncontrolledTooltip>
      <MapModal toggle={closeMap} isOpen={mapIsOpened} title={modalTitle} locations={locationsToShow} />
    </>
  );
};
