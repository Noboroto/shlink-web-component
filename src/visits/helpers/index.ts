import { countBy, groupBy } from '@shlinkio/data-manipulation';
import type { ShlinkOrphanVisit, ShlinkVisit, ShlinkVisitsParams } from '../../api-contract';
import type { ShortUrlIdentifier } from '../../short-urls/data';
import { domainMatches, shortUrlMatches } from '../../short-urls/helpers';
import { formatIsoDate, isBetween } from '../../utils/dates/helpers/date';
import type { DateRange } from '../../utils/dates/helpers/dateIntervals';
import type {
  CreateVisit,
  HighlightableProps,
  NormalizedOrphanVisit,
  NormalizedVisit,
  Stats,
  VisitsParams,
  VisitsQueryParams,
} from '../types';

export const isOrphanVisit = (visit: ShlinkVisit): visit is ShlinkOrphanVisit =>
  (visit as ShlinkOrphanVisit).visitedUrl !== undefined;

export const isNormalizedOrphanVisit = (visit: NormalizedVisit): visit is NormalizedOrphanVisit =>
  (visit as NormalizedOrphanVisit).visitedUrl !== undefined;

export interface GroupedNewVisits {
  orphanVisits: CreateVisit[];
  nonOrphanVisits: CreateVisit[];
}

export const groupNewVisitsByType = (createdVisits: CreateVisit[]): GroupedNewVisits => {
  const groupedVisits: Partial<GroupedNewVisits> = groupBy(
    createdVisits,
    (newVisit: CreateVisit) => (isOrphanVisit(newVisit.visit) ? 'orphanVisits' : 'nonOrphanVisits'),
  );
  return { orphanVisits: [], nonOrphanVisits: [], ...groupedVisits };
};

/**
 * Filters provided created visits by those matching a short URL and query
 */
export const filterCreatedVisitsByShortUrl = (
  createdVisits: CreateVisit[],
  { shortCode, domain }: ShortUrlIdentifier,
  { endDate, startDate }: VisitsQueryParams,
): CreateVisit[] => createdVisits.filter(
  ({ shortUrl, visit }) =>
    shortUrl && shortUrlMatches(shortUrl, shortCode, domain) && isBetween(visit.date, startDate, endDate),
);

/**
 * Filters provided created visits by those matching a domain and query
 */
export const filterCreatedVisitsByDomain = (
  createdVisits: CreateVisit[],
  domain: string,
  { endDate, startDate }: VisitsQueryParams,
): CreateVisit[] => createdVisits.filter(
  ({ shortUrl, visit }) => shortUrl && domainMatches(shortUrl, domain) && isBetween(visit.date, startDate, endDate),
);

/**
 * Filters provided created visits by those matching a domain and query
 */
export const filterCreatedVisitsByTag = (
  createdVisits: CreateVisit[],
  tag: string,
  { endDate, startDate }: VisitsQueryParams,
): CreateVisit[] => createdVisits.filter(
  ({ shortUrl, visit }) => shortUrl?.tags.includes(tag) && isBetween(visit.date, startDate, endDate),
);

export const highlightedVisitsToStats = <T extends NormalizedVisit>(
  highlightedVisits: T[],
  property: HighlightableProps<T>,
): Stats => countBy(highlightedVisits, (value: any) => value[property]);

export const toApiDateRange = (dateRange?: DateRange): Pick<ShlinkVisitsParams, 'startDate' | 'endDate'> => {
  const startDate = (dateRange?.startDate && formatIsoDate(dateRange?.startDate)) ?? undefined;
  const endDate = (dateRange?.endDate && formatIsoDate(dateRange?.endDate)) ?? undefined;

  return { startDate, endDate };
};

export const toApiParams = ({ page, itemsPerPage, filter, dateRange }: VisitsParams): ShlinkVisitsParams => {
  const { startDate, endDate } = toApiDateRange(dateRange);
  const excludeBots = filter?.excludeBots || undefined;

  return { page, itemsPerPage, startDate, endDate, excludeBots };
};
