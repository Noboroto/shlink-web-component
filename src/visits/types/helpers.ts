import type { ShlinkOrphanVisit, ShlinkVisit, ShlinkVisitsParams } from '../../api-contract';
import { formatIsoDate } from '../../utils/dates/helpers/date';
import { countBy, groupBy } from '../../utils/helpers/data';
import type { CreateVisit, NormalizedOrphanVisit, NormalizedVisit, Stats, VisitsParams } from './index';

export const isOrphanVisit = (visit: ShlinkVisit): visit is ShlinkOrphanVisit =>
  (visit as ShlinkOrphanVisit).visitedUrl !== undefined;

export const isNormalizedOrphanVisit = (visit: NormalizedVisit): visit is NormalizedOrphanVisit =>
  (visit as NormalizedOrphanVisit).visitedUrl !== undefined;

export interface GroupedNewVisits {
  orphanVisits: CreateVisit[];
  nonOrphanVisits: CreateVisit[];
}

export const groupNewVisitsByType = (createdVisits: CreateVisit[]): GroupedNewVisits => {
  const groupedVisits: Partial<GroupedNewVisits> = groupBy(
    (newVisit: CreateVisit) => (isOrphanVisit(newVisit.visit) ? 'orphanVisits' : 'nonOrphanVisits'),
    createdVisits,
  );
  return { orphanVisits: [], nonOrphanVisits: [], ...groupedVisits };
};

export type HighlightableProps<T extends NormalizedVisit> = T extends NormalizedOrphanVisit
  ? ('referer' | 'country' | 'city' | 'visitedUrl')
  : ('referer' | 'country' | 'city');

export const highlightedVisitsToStats = <T extends NormalizedVisit>(
  highlightedVisits: T[],
  property: HighlightableProps<T>,
): Stats => countBy((value: any) => value[property], highlightedVisits);

export const toApiParams = ({ page, itemsPerPage, filter, dateRange }: VisitsParams): ShlinkVisitsParams => {
  const startDate = (dateRange?.startDate && formatIsoDate(dateRange?.startDate)) ?? undefined;
  const endDate = (dateRange?.endDate && formatIsoDate(dateRange?.endDate)) ?? undefined;
  const excludeBots = filter?.excludeBots || undefined;

  return { page, itemsPerPage, startDate, endDate, excludeBots };
};
