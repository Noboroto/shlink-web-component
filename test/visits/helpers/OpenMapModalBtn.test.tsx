import { screen, waitFor } from '@testing-library/react';
import { fromPartial } from '@total-typescript/shoehorn';
import { OpenMapModalBtn } from '../../../src/visits/helpers/OpenMapModalBtn';
import type { CityStats } from '../../../src/visits/types';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<OpenMapModalBtn />', () => {
  const title = 'Foo';
  const locations: CityStats[] = [
    fromPartial({ cityName: 'foo', count: 30, latLong: [5, 5] }),
    fromPartial({ cityName: 'bar', count: 45, latLong: [88, 88] }),
  ];
  const setUp = (activeCities?: string[]) => renderWithEvents(
    <OpenMapModalBtn modalTitle={title} locations={locations} activeCities={activeCities} />,
  );

  it('renders tooltip on button hover and opens modal on click', async () => {
    const { user } = setUp();

    expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();

    await user.click(screen.getByRole('button'));
    await waitFor(() => expect(screen.getByRole('tooltip')).toBeInTheDocument());
    await waitFor(() => expect(screen.getByRole('dialog')).toBeInTheDocument());
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
  });

  it('opens dropdown instead of modal when a list of active cities has been provided', async () => {
    const { user } = setUp(['bar']);

    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();

    await user.click(screen.getByRole('button'));

    await waitFor(() => expect(screen.getByRole('menu')).toBeInTheDocument());
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it.each([
    ['Show all locations'],
    ['Show locations in current page'],
  ])('filters out non-active cities from list of locations', async (name) => {
    const { user } = setUp(['bar']);

    await user.click(screen.getByRole('button'));
    await user.click(screen.getByRole('menuitem', { name }));

    expect(await screen.findByRole('dialog')).toMatchSnapshot();
  });
});
