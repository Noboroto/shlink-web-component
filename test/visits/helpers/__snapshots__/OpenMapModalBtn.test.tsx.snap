// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OpenMapModalBtn /> > filters out non-active cities from list of locations 1`] = `
<div
  class="modal fade"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog map-modal__modal"
    role="document"
  >
    <div
      class="modal-content map-modal__modal-content"
    >
      <div
        class="map-modal__modal-body modal-body"
      >
        <h3
          class="map-modal__modal-title"
        >
          Foo
          <button
            aria-label="Close"
            class="btn-close float-end"
            type="button"
          />
        </h3>
        <div
          class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
          style="position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            >
              <div
                class="leaflet-layer "
                style="z-index: 1;"
              >
                <div
                  class="leaflet-tile-container leaflet-zoom-animated"
                  style="z-index: 18; left: 0px; top: 0px;"
                >
                  <img
                    alt=""
                    class="leaflet-tile"
                    src="https://a.tile.openstreetmap.org/0/0/0.png"
                    style="width: 256px; height: 256px; left: -161px; top: -62px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-shadow-pane"
            >
              <img
                alt=""
                class="leaflet-marker-shadow leaflet-zoom-hide"
                src="marker-shadow.png"
                style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: -29px; top: 62px;"
              />
              <img
                alt=""
                class="leaflet-marker-shadow leaflet-zoom-hide"
                src="marker-shadow.png"
                style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 30px; top: -62px;"
              />
            </div>
            <div
              class="leaflet-pane leaflet-marker-pane"
            >
              <img
                alt="Marker"
                class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
                role="button"
                src="marker-icon.png"
                style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: -29px; top: 62px; z-index: 62;"
                tabindex="0"
              />
              <img
                alt="Marker"
                class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
                role="button"
                src="marker-icon.png"
                style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 30px; top: -62px; z-index: -62;"
                tabindex="0"
              />
            </div>
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            />
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-disabled="false"
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  <span
                    aria-hidden="true"
                  >
                    +
                  </span>
                </a>
                <a
                  aria-disabled="true"
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out leaflet-disabled"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  <span
                    aria-hidden="true"
                  >
                    −
                  </span>
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JavaScript library for interactive maps"
                >
                  Leaflet
                </a>
                 
                <span
                  aria-hidden="true"
                >
                  |
                </span>
                 © 
                <a
                  href="https://osm.org/copyright"
                >
                  OpenStreetMap
                </a>
                 contributors
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<OpenMapModalBtn /> > filters out non-active cities from list of locations 2`] = `
<div
  class="modal fade"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog map-modal__modal"
    role="document"
  >
    <div
      class="modal-content map-modal__modal-content"
    >
      <div
        class="map-modal__modal-body modal-body"
      >
        <h3
          class="map-modal__modal-title"
        >
          Foo
          <button
            aria-label="Close"
            class="btn-close float-end"
            type="button"
          />
        </h3>
        <div
          class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
          style="position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            >
              <div
                class="leaflet-layer "
                style="z-index: 1;"
              >
                <div
                  class="leaflet-tile-container leaflet-zoom-animated"
                  style="z-index: 18; left: 0px; top: 0px;"
                >
                  <img
                    alt=""
                    class="leaflet-tile"
                    src="https://a.tile.openstreetmap.org/10/762/0.png"
                    style="width: 256px; height: 256px; left: -80px; top: 0px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-shadow-pane"
            >
              <img
                alt=""
                class="leaflet-marker-shadow leaflet-zoom-hide"
                src="marker-shadow.png"
                style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 0px; top: 0px;"
              />
            </div>
            <div
              class="leaflet-pane leaflet-marker-pane"
            >
              <img
                alt="Marker"
                class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
                role="button"
                src="marker-icon.png"
                style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 0px; top: 0px; z-index: 0;"
                tabindex="0"
              />
            </div>
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            />
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-disabled="false"
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  <span
                    aria-hidden="true"
                  >
                    +
                  </span>
                </a>
                <a
                  aria-disabled="false"
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  <span
                    aria-hidden="true"
                  >
                    −
                  </span>
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JavaScript library for interactive maps"
                >
                  Leaflet
                </a>
                 
                <span
                  aria-hidden="true"
                >
                  |
                </span>
                 © 
                <a
                  href="https://osm.org/copyright"
                >
                  OpenStreetMap
                </a>
                 contributors
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
